File: food-finder-backend/src/routes/restaurant.py
import requests
import os
from flask import Blueprint, request, jsonify

restaurant_bp = Blueprint('restaurant', __name__)
    
# 음식 취향 키워드 매핑
FOOD_KEYWORDS = {
    'spicy': ['매운', '떡볶이', '김치찌개', '마라탕', '불닭'], 'mild': ['순한', '된장찌개', '미역국', '백반', '정식'],
    'korean': ['한식', '김치', '불고기', '갈비', '비빔밥'], 'western': ['양식', '파스타', '피자', '스테이크', '햄버거'],
    'rice': ['밥', '덮밥', '볶음밥', '비빔밥', '정식'], 'noodle': ['면', '라면', '냉면', '파스타', '우동'],
    'meat': ['고기', '삼겹살', '갈비', '치킨', '스테이크'], 'seafood': ['해산물', '회', '조개', '새우', '게'],
    'hot': ['뜨거운', '찌개', '국물', '탕', '전골'], 'cold': ['차가운', '냉면', '회', '샐러드', '아이스크림'],
    'salty': ['짠', '젓갈', '김치', '라면', '치킨'], 'sweet': ['단', '디저트', '케이크', '아이스크림', '과일'],
    'traditional': ['전통', '한정식', '백반', '정통', '옛날'], 'modern': ['모던', '퓨전', '신메뉴', '트렌드', '새로운'],
    'alone': ['혼밥', '1인분', '간단한', '가벼운', '테이크아웃'], 'group': ['단체', '회식', '모임', '가족', '여럿이']
}

def perform_kakao_search(params, api_key):
    """카카오 API 검색을 수행하고 결과를 반환하는 헬퍼 함수"""
    url = 'https://dapi.kakao.com/v2/local/search/keyword.json'
    headers = {'Authorization': f'KakaoAK {api_key}'}
    print(f"🔍 카카오 API 요청: {params}") # 디버깅을 위해 요청 파라미터 출력
    response = requests.get(url, headers=headers, params=params)
    response.raise_for_status()
    return response.json()

@restaurant_bp.route('/search', methods=['POST'])
def search_restaurants():
    """사용자 답변 기반으로 맛집 검색. 결과 없으면 키워드를 줄여 재검색합니다."""
    try:
        kakao_api_key = os.environ.get('KAKAO_REST_API_KEY')
        if not kakao_api_key:
            return jsonify({'success': False, 'message': '서버에 API 키가 설정되지 않았습니다.'}), 500
        
        data = request.get_json(force=True)
        answers = data.get('answers', [])
        location = data.get('location', {})
        
        # 1. 키워드 생성
        keywords = [kw for answer in answers if answer in FOOD_KEYWORDS for kw in FOOD_KEYWORDS[answer]]
        unique_keywords = list(set(keywords))
        
        params = {'size': 10, 'category_group_code': 'FD6'}
        if location.get('latitude') and location.get('longitude'):
            params.update({
                'x': str(location['longitude']),
                'y': str(location['latitude']),
                'radius': 2000, # 검색 반경 2km
                'sort': 'distance'
            })
        
        # 2. 1단계 검색: 구체적인 키워드 (최대 2개)로 검색
        params['query'] = ' '.join(unique_keywords[:2]) if unique_keywords else '맛집'
        result = perform_kakao_search(params, kakao_api_key)
        
        # 3. 2단계 검색: 1단계 결과가 없으면, 가장 핵심적인 키워드 1개로 재검색
        if not result.get('documents') and len(unique_keywords) > 1:
            params['query'] = unique_keywords[0]
            result = perform_kakao_search(params, kakao_api_key)

        # 4. 최종 결과 포맷팅
        restaurants = [
            {
                'id': item.get('id'), 'name': item.get('place_name'),
                'category': item.get('category_name'), 'address': item.get('address_name'),
                'phone': item.get('phone'),
                'distance': f"{item.get('distance')}m" if item.get('distance') else '알 수 없음',
                'url': item.get('place_url'),
                'x': item.get('x'), 'y': item.get('y'),
            }
            for item in result.get('documents', [])
        ]
        
        return jsonify({'success': True, 'restaurants': restaurants})
            
    except requests.exceptions.HTTPError as err:
        error_message = err.response.json().get("message", "알 수 없는 API 오류")
        return jsonify({'success': False, 'message': f'API 호출 실패: {error_message}'}), err.response.status_code
    except Exception as e:
        return jsonify({'success': False, 'message': f'서버 오류: {str(e)}'}), 500
# 상태 확인용 API (변경 없음)
@restaurant_bp.route('/health', methods=['GET'])
def health_check():
    """API 상태 확인"""
    return jsonify({'status': 'healthy', 'message': '맛집 검색 API가 정상 작동 중입니다.'})

File: food-finder-backend-node/routes/restaurant.js
// routes/restaurant.js
const express = require('express');
const router = express.Router();
const axios = require('axios'); // 이미지 프록시용

// 키워드 정의
const FOOD_KEYWORDS = {
    'korean':   { type: 'cuisine', keywords: ['한식'] },
    'western':  { type: 'cuisine', keywords: ['양식', '이탈리안', '브런치'] },
    'chinese':  { type: 'cuisine', keywords: ['중식', '중화요리'] },
    'japanese': { type: 'cuisine', keywords: ['일식', '초밥', '이자카야'] },
    'sweet':    { type: 'cuisine', keywords: ['디저트', '카페'] },

    'meat':     { type: 'ingredient', keywords: ['고기', '구이', '스테이크'] },
    'seafood':  { type: 'ingredient', keywords: ['해산물', '회', '매운탕'] },
    'rice':     { type: 'ingredient', keywords: ['밥집', '덮밥', '정식'] },
    'noodle':   { type: 'ingredient', keywords: ['면요리', '국수', '파스타', '짬뽕'] },

    'spicy':    { type: 'flavor', keywords: ['매운', '얼큰한'] },
    'mild':     { type: 'flavor', keywords: ['순한', '담백한'] },
    'hot':      { type: 'flavor', keywords: ['따뜻한', '국물', '전골', '찌개'] },
    'cold':     { type: 'flavor', keywords: ['시원한', '냉면'] },
    'salty':    { type: 'flavor', keywords: ['짭짤한'] },

    'alcohol':  { type: 'style', keywords: ['술집', '포차'] },
    'alone':    { type: 'style', keywords: ['혼밥'] },
    'group':    { type: 'style', keywords: ['단체'] },
    'modern':   { type: 'style', keywords: ['분위기 좋은'] },
    'traditional': { type: 'style', keywords: ['전통', '노포'] },

    'near':     { type: 'distance', keywords: [] }, 
    'far':      { type: 'distance', keywords: [] }
};

// [핵심 변경] 구글 검색 안 함! 검색어(String)만 생성해서 응답함.
router.post('/search', (req, res) => {
    try {
        const { answers } = req.body;
        const categorized = { cuisine: [], ingredient: [], flavor: [], style: [], distance: [] };
        
        answers.forEach(answer => {
            if (!answer) return;
            const keywordInfo = FOOD_KEYWORDS[answer];
            if (keywordInfo) {
                if (!categorized[keywordInfo.type]) categorized[keywordInfo.type] = [];
                categorized[keywordInfo.type].push(...keywordInfo.keywords);
                
                // [추가] 거리 타입은 별도로 저장 (키워드 자체를 저장)
                if (keywordInfo.type === 'distance') {
                    categorized.distance.push(answer);
                }
            }
        });

        // 검색어 조합
        const queryParts = [];
        
        const isDessert = categorized.cuisine.includes('디저트') || categorized.cuisine.includes('카페');

        if (categorized.cuisine.length > 0) queryParts.push(categorized.cuisine[0]);

        if (!isDessert) {
            if (categorized.ingredient.length > 0) queryParts.push(categorized.ingredient[0]);
            if (categorized.flavor.length > 0) queryParts.push(categorized.flavor[0]);
        } else {
            if (categorized.style && categorized.style.length > 0) queryParts.push(categorized.style[0]);
        }

        // 술/분위기 키워드는 상황에 따라 뒤에 붙임
        if (categorized.style.includes('술집') || categorized.style.includes('포차')) {
             queryParts.push(categorized.style[0]);
        }

        let finalQuery = queryParts.join(' ');
        
        // 검색어가 너무 휑하면(예: 다 null이라서 빈 문자열이면) 기본값 설정
        if (!finalQuery.trim()) {
            finalQuery = "맛집"; 
        } else if (!finalQuery.includes('맛집') && !finalQuery.includes('카페') && !finalQuery.includes('술집')) {
             finalQuery += ' 맛집';
        }
        let searchRadius = 50000;
        if (categorized.distance.includes('near')) {
            searchRadius = 3000; // '근처' 선택 시 3km 이내로 제한
            console.log("📍 거리 제한 적용: 가까운 곳 (3km)");
        } else if (categorized.distance.includes('far')) {
            searchRadius = 80000; // '멀리' 선택 시 80km까지 확장
            console.log("📍 거리 제한 적용: 멀리 (80km)");
        }
        console.log(`✅ 전달 검색어: "${finalQuery}", 반경: ${searchRadius}m`);
        res.json({ success: true, query: finalQuery, radius: searchRadius });

    } catch (error) {
        console.error("검색어 생성 오류:", error);
        res.status(500).json({ success: false, message: '검색어 생성 실패' });
    }
});

// 이미지 프록시 (유지)
router.get('/image/:photo_reference', async (req, res) => {
    try {
        const photoReference = req.params.photo_reference;
        const googleApiKey = process.env.GOOGLE_MAPS_API_KEY;

        if (!photoReference || !googleApiKey) {
            return res.status(400).send('Bad Request');
        }

        const url = 'https://maps.googleapis.com/maps/api/place/photo';
        const response = await axios.get(url, {
            params: {
                maxwidth: 400,
                photoreference: photoReference,
                key: googleApiKey
            },
            responseType: 'stream'
        });
        response.data.pipe(res);
    } catch (error) {
        res.status(404).send('Image not found');
    }
});

module.exports = router;